src/contracts/core/StakeRootCompendiumStorage.sol

```diff
@@ -0,0 +1,90 @@
+import "@openzeppelin/contracts/utils/structs/EnumerableMap.sol";
+import "../interfaces/IAVSDirectory.sol";
+import "../interfaces/IDelegationManager.sol";
+import "../interfaces/IAllocationManager.sol";
+import "../interfaces/IStakeRootCompendium.sol";
+import "../libraries/Snapshots.sol";
+
+abstract contract StakeRootCompendiumStorage is IStakeRootCompendium, OwnableUpgradeable {
+    using Snapshots for Snapshots.History;
+    using EnumerableMap for EnumerableMap.AddressToUintMap;
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    /// @notice the delegation manager contract
+    IDelegationManager public immutable delegationManager;
+    /// @notice the AVS directory contract
+    IAVSDirectory public immutable avsDirectory;
+    /// @notice the allocation manager contract
+    IAllocationManager public immutable allocationManager;
+
+    /// @notice the minimum balance that must be maintained for an operatorSet
+    /// @dev this balance compensates gas costs to deregister an operatorSet
+    uint256 public immutable MIN_BALANCE_THRESHOLD;
+
+    /// @notice the placeholder index used for operator sets that are removed from the StakeTree
+    uint32 public constant REMOVED_INDEX = type(uint32).max;
+
+    /// @notice the minimum number of proofs that an operatorSet's deposit balance needs to cover and
+    /// the number of proofs they must pay for since their latest reconfiguration
+    /// @dev this prevents de-registering an operatorSet immediately after reconfiguring
+    uint256 public immutable MIN_PREPAID_PROOFS;
+
+    /// @notice the total number of strategies among all operator sets (with duplicates)
+    uint256 public totalStrategies;
+
+    /// @notice the verifier contract that will be used to verify snark proofs
+    address public immutable verifier;
+    /// @notice the id of the program being verified when roots are posted
+    bytes32 public immutable imageId;
+
+    /// @notice the address allowed to confirm roots
+    address public rootConfirmer;
+
+    /// @notice list of operator sets that have been configured to be in the StakeTree
+    OperatorSet[] public operatorSets;
+    /// @notice the total charge for a proofs at a certain time depending on the number of strategies
+    Snapshots.History internal chargePerProof;
+    /// @dev Contains cumulative charges for operator sets, strategies, and max total charge.
+    ChargeParams public chargeParams;
+    /// @dev Contains info about cumulative charges.
+    CumulativeChargeParams public cumulativeChargeParams;
+    /// @notice deposit balance to be deducted for operatorSets
+    mapping(address => mapping(uint32 => DepositInfo)) public depositInfos;
+    /// @notice the extraData for each operatorSet
+    mapping(address => mapping(uint32 => bytes32)) internal operatorSetExtraDatas;
+    /// @notice map from operator set to a trace of their index over time
+    mapping(address => mapping(uint32 => Snapshots.History)) internal operatorSetToIndex;
```

Comment:
the part of the picture that doesn't exist now is that, when operatorSets prove their opsetRoot against the stakeRoot, they know what index to prove against the stakeRoot

alternatively, we could redefine `opsetRoot_new = hash(opset, opsetRoot_curr)` so that it is inherent in the leaf, similar to rewards

wdyt?